src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseCodeReviewFeedbackService.java

```diff
@@ -0,0 +1,227 @@
+                                feedbackText = String.format(NON_GRADED_FEEDBACK_SUGGESTION + "File %s at lines %d-%d", individualFeedbackItem.filePath(),
+                                        individualFeedbackItem.lineStart(), individualFeedbackItem.lineEnd());
+                            }
+                            else {
+                                feedbackText = String.format(NON_GRADED_FEEDBACK_SUGGESTION + "File %s at line %d", individualFeedbackItem.filePath(),
+                                        individualFeedbackItem.lineStart());
+                            }
+                            feedback.setReference(String.format("file:%s_line:%d", individualFeedbackItem.filePath(), individualFeedbackItem.lineStart()));
+                        }
+                        else {
+                            feedbackText = String.format(NON_GRADED_FEEDBACK_SUGGESTION + "File %s", individualFeedbackItem.filePath());
+                        }
+                        feedback.setText(feedbackText);
+                        feedback.setDetailText(individualFeedbackItem.description());
+                        feedback.setHasLongFeedbackText(false);
+                        feedback.setType(FeedbackType.AUTOMATIC);
+                        feedback.setCredits(0.0);
+                        return feedback;
+                    }).toList();
+
+            automaticResult.setSuccessful(true);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+
+            this.resultService.storeFeedbackInResult(automaticResult, feedbacks, true);
+
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        catch (Exception e) {
+            log.error("Could not generate feedback", e);
+            automaticResult.setSuccessful(false);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+            this.resultRepository.save(automaticResult);
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        finally {
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation(participation);
+            participation.setIndividualDueDate(null);
+            this.programmingExerciseStudentParticipationRepository.save(participation);
+        }
+    }
+
+    /**
+     * Sets an individual due date for a participation, locks the repository,
+     * and invalidates previous results to prepare for new feedback.
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private ProgrammingExerciseStudentParticipation setIndividualDueDateAndLockRepository(ProgrammingExerciseStudentParticipation participation,
+            ProgrammingExercise programmingExercise, boolean invalidatePreviousResults) {
+        var currentDate = now();
```

Comment:
does the introduction of this variable improve readability? Does it have another use?

Reply:
inlined